design-docs/incremental-build.md

```diff
@@ -243,3 +243,115 @@ When the `TaskHistory` gets persisted, it adds the current task execution to the
+         */
+        void registerStrategy(BuildOutputCleanupStrategy strategy);
+        
+        /**
+         * Returns all registered clean up strategies.
+         */
+        List<BuildOutputCleanupStrategy> getStrategies();
+        
+        /**
+         * Registers outputs to be cleaned up.
+         */
+        void registerOutputs(File... outputs);
+
+        /**
+         * Registers outputs to be cleaned up.
+         */
+        void registerOutputs(List<File> outputs);
+        
+        /**
+         * Returns all registed outputs.
+         */
+        List<File> getOutputs();
+    }
+
+- Provide an implementation for a clean up strategy that identifies if task history is available.
+    - Create an instance of the implementation in Gradle's global service registry.
+    - Persist the Gradle version to a cross-version file [similar to what we have for `buildSrc`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java#L103).
+    - Only initiate a clean if one of the following conditions is encountered:
+        - If the cross-version file does not exist.
+        - If the current Gradle version used by the build is different from the one persisted in the cross-version file.
+    - Stale files produced by older versions of Gradle (that do not know about the registry implementation) are out of scope.
+
+<!-- -->
+
+    public class PersistedGradleVersionCleanupStrategy implements BuildOutputCleanupStrategy {
+        @Override
+        public boolean needsCleanup(Gradle gradle) {
+            // identify existing task history
+        }
+    }
+
+- Clean up registry concept applies to output of Java compilation.
+    - Register the task history clean up implementation with the clean up registry in the `JavaPlugin`. 
+    - As output define outputs of the `main` and `test` source set.
+    - Add classes output directory via `sourceSet.getOutput().getClassesDir()`.
+    - Add resources output directory via `sourceSet.getOutput().getResourcesDir()`.
+- Delete outputs if needed.
+    - Retrieve all clean up strategies as one of the last things during the configuration phase.
+    - Delete outputs for a strategy if needed. If output does not exist then skip. Render a message with log level `quiet` to the console that indicates operation.
+    - The delete operation is not represented by a task in task graph.
+    - Failure to delete an output (e.g. due to file locking) will render a helpful warning message but not fail the build.
+- Cleaning the output works the same way when executing in parallel mode (via `--parallel` command line option or setting in `gradle.properties`).
+- As dogfooding exercise also use the registry for cleaning up `buildSrc` if task history is out-of-date.
+- Out of scope for this story are the following aspects:
+    - Do not apply the concept to outputs of custom source sets created by a user.
```

Comment:
Maybe I'm missing something, but right now I don't see this becoming a public API. As far as I understand the problem, this is a workaround for the fact that we have tasks with overlapping output dirs. If that wasn't the case, the `JavaCompile` task could just clean its output dir when it is not incremental. So I see this "stale output registry" more as a workaround until we clean up that overlap situation.